96 bounds predprob

audreyyeoCH opened the pull request:
closes #96

audreyyeoCH commented:
I found some test cases failing and am working on it before I feel confident to merge. This test case seems to be unrelated to the upper boundary of density in `predprob()` being 1 and not 1 + Machine$double.eps. Still working on it...

audreyyeoCH commented:
> I found some test cases failing and am working on it before I feel confident to request PR. This test case seems to be unrelated to the upper boundary of density in `predprob()` being 1 and not 1 + Machine$double.eps. Still working on it...

 I solved it! The test case passed when`predprob()` used to check these results.

audreyyeoCH commented:
I removed the assignment of default weight, and its assertion at the start of the user facing function (`boundsPredprob()`) because it is already assigned and asserted in the helper function of `predprob()`:

`if (missing(weights)) {
weights <- rep(1, nrow(parE))
 }
assert_numeric(weights, min.len = 0, len = nrow(par), finite = TRUE)`

if this is still wanted, I would suggest the following to avoid incorrect results:

`
parE <- t(parE)
if (missing(weights)) {
weights <- rep(1, nrow(parE))
 }
assert_numeric(weights, min.len = 0, len = nrow(par), finite = TRUE)
parE <- t(parE) # repeated so it doesn't give incorrect results as predprob already sets the parE up in the right dimension`

audreyyeoCH commented:
Hi @danielinteractive could you kindly review please ?

danielinteractive commented on examples/plotBounds.R:
```diff
@@ -1,9 +1 @@
 # examples
-plotBounds(boundsPredprob(
```
@audreyyeoCH can you please first merge in `main` into this feature branch? Because currently these examples are deleted, but they should be there (because https://github.com/Genentech/phase1b/blob/main/examples/plotBounds.R has them)

danielinteractive replied:
This file still has "deleted" contents, but should not, because these examples were just added in the previous PR to `main`?

danielinteractive commented on R/boundsPredprob.R:
```diff
@@ -1,74 +1,86 @@
-#' Decision cutpoints for boundary (based on predictive probability)
+#' Decision cutpoints for boundary (based on predictive probability) for Decision 1 rule.
```
Since you mention Decision 1 rule, better have a section in `@details` which references the definition of the Decision 1 rule

danielinteractive replied:
Has this been added already?

danielinteractive replied:
@audreyyeoCH I don't see the details yet?

danielinteractive commented on R/boundsPredprob.R:
```diff
@@ -1,74 +1,86 @@
 #' Efficacy boundary: find minimum x (xU) where
-#' Pr(Pr(P > p | x, Y) >= tT | x) > phiU,
+#' Pr(Pr(P > p0 | x, Y, a, b) >= tT | x) >= phiU,
 #' Futility boundary: find maximum x (xL) where
-#' Pr(Pr(P > p | x, Y) >= tT | x) < phiL
+#' Pr(Pr(P > p0 | x, Y, a, b) >= tT | x) =< phiL
```
do we always include `a` and `b` on the condition side of these probabilities? It is not really precise because it might also be a mixture of beta distributions as a prior (i.e. not just a and b as prior parameters), so I am not sure this is needed/useful

danielinteractive commented on R/boundsPredprob.R:
```diff
@@ -1,74 +1,86 @@
+#' - `xU` : the minimal number of responses that meet the efficacy threshold.
+#' - `pU` : response rate corresponding to `xU`.
+#' - `predU` : predictive probability corresponding to `xU`
+#' - `postL`: posterior probability corresponding to `xU`.
```
```suggestion
#' - `postU`: posterior probability corresponding to `xU`.
```

danielinteractive commented on R/boundsPredprob.R:
```diff
@@ -1,74 +1,86 @@
+#' boundary based on the following rules:
 #' Efficacy boundary: find minimum x (xU) where
-#' Pr(Pr(P > p | x, Y) >= tT | x) > phiU,
+#' Pr(Pr(P > p0 | x, Y, a, b) >= tT | x) >= phiU,
```
I remember we had `RR` in the other PR instead of `P` here, would be good to make this consistent throughout this function's documentation

danielinteractive commented on tests/testthat/test-boundsPredprob.R:
```diff
@@ -0,0 +1,100 @@
+      postU = c(0.9496, 0.9569, 0.9254, 0.8177),
+      LciU = c(0.15, 0.1773, 0.1663, 0.1424)
+    )
+  )
```
please also check `predL` and `predU` here

audreyyeoCH commented:
Am trying to reply to above comment : "This file still has "deleted" contents, but should not, because these examples were just added in the previous PR to main?"

I deleted them again, because they will cause errors as I updated `boundPredprob()`

audreyyeoCH commented:
Hi @danielinteractive the R CMD checks passed, and Doug is looking into the Roxygen error (around `glue`). I'll request a review if you had a moment : )

danielinteractive reviewed (commented):
Hi @audreyyeoCH , the file `examples/plotBounds.R` is still modified in this PR, but should not. Or am I missing anything?

audreyyeoCH commented:
> Hi @audreyyeoCH , the file `examples/plotBounds.R` is still modified in this PR, but should not. Or am I missing anything?

Yes it should be modified, because `plotBounds(boundsPredprob(old function stuff))` should cause R CMD check errors. I indeed modified `examples/plotBounds.R`such that it should be empty. Let me know if I answered your question. I get that we should not modify other stuff while working on this function, but `plotBounds` will be modified soon and it's only the example that has been removed, but the file is still there.

danielinteractive commented:
> > Hi @audreyyeoCH , the file `examples/plotBounds.R` is still modified in this PR, but should not. Or am I missing anything?
> 
> Yes it should be modified, because `plotBounds(boundsPredprob(old function stuff))` should cause R CMD check errors. I indeed modified `examples/plotBounds.R`such that it should be empty. Let me know if I answered your question. I get that we should not modify other stuff while working on this function, but `plotBounds` will cause check errors that won't be allowed to merge

Aha I see, thank you - but then would it be possible to modify the examples for `plotBounds` accordingly to the updates of `boundsPredprob` such that they work? Because we want to keep the functionality (in this case having examples that work for `plotBounds` together with `boundsPredprob`)

audreyyeoCH commented:
> > > Hi @audreyyeoCH , the file `examples/plotBounds.R` is still modified in this PR, but should not. Or am I missing anything?
> > 
> > 
> > Yes it should be modified, because `plotBounds(boundsPredprob(old function stuff))` should cause R CMD check errors. I indeed modified `examples/plotBounds.R`such that it should be empty. Let me know if I answered your question. I get that we should not modify other stuff while working on this function, but `plotBounds` will cause check errors that won't be allowed to merge
> 
> Aha I see, thank you - but then would it be possible to modify the examples for `plotBounds` accordingly to the updates of `boundsPredprob` such that they work? Because we want to keep the functionality (in this case having examples that work for `plotBounds` together with `boundsPredprob`)

Thanks @danielinteractive, I totally want to do that but I'd like to use the `ggplot2` grammar instead of the `graphics` one to keep it consistent throughout the package e.g. more people know `ggplot2`, that's just one idea. If I modified the  examples for `plotBounds` then want to switch to `ggplot2` grammar, it would be doing double the polishing, let me know your thoughts!

danielinteractive commented:
Hm I see. I think only some names change in the output of `boundsPredprob` right? In that case could just  have a temporary object assigned the result of it, and then rename and pass to `plotBounds`. I would recommend keeping that example such that later it will be easy to compare when you refactor `plotBounds`. So a little bit extra code now is worth it.

audreyyeoCH commented:
> Hm I see. I think only some names change in the output of `boundsPredprob` right? In that case could just have a temporary object assigned the result of it, and then rename and pass to `plotBounds`. I would recommend keeping that example such that later it will be easy to compare when you refactor `plotBounds`. So a little bit extra code now is worth it.

ahh true that - thanks! I added it back now... and `plotBounds(boundsPOSTprob))` 😆 , thanks for your review!

danielinteractive commented on examples/boundsPredprob.R:
```diff
@@ -1,12 +1,33 @@
+# Final efficacy decision if more than 80% probability to be above 20% ORR,
+# Final futility decision otherwise.
+# Interim efficacy decision if more than 90% predictive probability reach this or
+# Efficacy look Pr(Pr(P > p0 | x, Y, a, b) >= tT | x) >= phiU,
+# Interim futility decision if less than 10% predictive probability or
+# Futility look Pr(Pr(P > p0 | x, Y, a, b) >= tT | x) =< phiL
```
also adjust notation here with `RR` etc.

danielinteractive commented on examples/ocPredprobDist.R:
```diff
@@ -41,10 +41,14 @@ result$oc
-# - Final look for Futility: Pr( response rate + deltaF < 25% ) < 60% or P(response rate + deltaF > p0) < tT
-# - Interim look for Efficacy: Pr( success at final ) > 80% or P(success at final) > phiU
-# - Interim look for Futility: Pr( failure at final ) < 20% or P(success at final) < phiL
+# - Final look for Efficacy: Pr( response rate + deltaE > 25% ) > 60% or
```
better use `RR` instead of `response rate` in the formulas in general (then it is easier to read as well)

danielinteractive reviewed (commented):
Thanks @audreyyeoCH , nice work, please see comments below

danielinteractive commented on examples/boundsPredprob.R:
```diff
@@ -2,9 +2,9 @@
+# Efficacy look Pr(Pr(RR > p0 | x, Y) >= tT | x) >= phiU,
 # Interim futility decision if less than 10% predictive probability or
-# Futility look Pr(Pr(P > p0 | x, Y, a, b) >= tT | x) =< phiL
+# Futility look Pr(Pr(RR) > p0 | x, Y) >= tT | x) =< phiL
```
```suggestion
# Futility look Pr(Pr(RR > p0 | x, Y) >= tT | x) =< phiL
```

audreyyeoCH commented:
ok so I had some debugging to do...

the `t(parE)` seems to be duplicating the efforts of helper functions `postprob()` and `predprob()` so I removed it. Which created downward problem with ìfmissing(weight)`, so I fixed that too.

the tests are a little more improved too. The first test tests if weights are there, the second builds on what if the weights are there and there are mixed priors.

danielinteractive reviewed (commented):
Thanks @audreyyeoCH , almost there!
Btw it would be good to soon come back to a state where only 1 PR is in the works, and the PR is as small as possible. That way, it will be quicker to finish and get to a state where all tests pass etc. again.

audreyyeoCH commented:
> Thanks @audreyyeoCH , almost there! Btw it would be good to soon come back to a state where only 1 PR is in the works, and the PR is as small as possible. That way, it will be quicker to finish and get to a state where all tests pass etc. again.

Yes for sure!

danielinteractive commented on R/boundsPredprob.R:
```diff
@@ -52,7 +52,7 @@ boundsPredprob <- function(looks, Nmax = max(looks), p0, tT, phiL, phiU, parE =
     weights <- rep(1, nrow(t(parE)))
   }
-  assert_numeric(weights, min.len = 0, len = nrow(par), finite = TRUE)
+  assert_numeric(weights, min.len = 0, len = nrow(parE), finite = TRUE)
```
If `parE` is passed as a vector then this would fail, I guess we also need `t(parE)`, as in line 53?
